### PR TITLE
Fix newly created token through the precompile not accessible through ERC20/721 redirect functions

### LINKED ISSUE #3150: Newly created token through the precompile not accessible through ERC20/721 redirect functions
### Description

I tried using a newly created token through the precompile not accessible through ERC20/721 redirect functions. Unfortunately loading the token through the Stacked World Updaters results in `INVALID_SOLIDITY_ADDRESS`. 
Need to move logic recognizing token accounts and returning the redirect bytecode to Stacked updaters as well (currently it's only applied to HederaWorldState.Udater)

### Steps to reproduce

- I created a token through the smart contract service using the new Token Create Precompile function.
- I tried using the address of the newly created token for a ERC721 call
- I got a `INVALID_SOLIDITY_ADDRESS` when loading the account

### Additional context

_No response_

### Hedera network

other

### Version

local build latest master

### Operating system

_No response_

### PR DESCRIPTION
**Pull request by dimitar-dinev**
…functions

Signed-off-by: Dimitar Dinev <mitkojc@gmail.com>

**Description**:

**Related issue(s)**:

Fixes #3150 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/store/contracts/HederaStackedWorldStateUpdater.java [hunk 1]
```diff
@@ -180,6 +219,19 @@ private boolean isMissingTarget(final Address alias) {
...
+			parentUpdater = parentUpdater.get().parentUpdater();
+		}
+		if (parentUpdater.isEmpty()) {
+			throw new InvalidTransactionException(ResponseCodeEnum.FAIL_INVALID);
+		}
+
+		final var topLevelUpdater = (HederaWorldState.Updater) parentUpdater.get();
+		return (HederaWorldState) topLevelUpdater.wrappedWorldView();
+	}
+
```
Comment (dimitar-dinev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/HederaStackedWorldStateUpdater.java:234]: Is this the best approach we can take?

Another would be to extract both `WorldStateAccount` and `WorldStateTokenAccount` from `HederaWorldState` into their own separate classes, which would be used both here and in `HederaWorldState`.

File: hedera-node/src/main/java/com/hedera/services/store/contracts/AbstractLedgerWorldUpdater.java [hunk 2]
```diff
@@ -144,9 +144,7 @@ public EvmAccount createAccount(final Address addressOrAlias, final long nonce,
 	}
 
 	@Override
-	public Account get(final Address addressOrAlias) {
-		final var address = aliases().resolveForEvm(addressOrAlias);
```
Comment (stoqnkpL) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/AbstractLedgerWorldUpdater.java:148]: Why do we need to remove this line? I think we should keep it or we may brake other logic like create2

File: hedera-node/src/main/java/com/hedera/services/store/contracts/AbstractLedgerWorldUpdater.java [hunk 3]
```diff
@@ -156,9 +154,7 @@ public Account get(final Address addressOrAlias) {
 	}
 
 	@Override
-	public EvmAccount getAccount(final Address addressOrAlias) {
-		final var address = aliases().resolveForEvm(addressOrAlias);
```
Comment (stoqnkpL) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/AbstractLedgerWorldUpdater.java:160]: Why do we need to remove this line? I think we should keep it or we may brake other logic like create2

File: hedera-node/src/main/java/com/hedera/services/store/contracts/AbstractLedgerWorldUpdater.java (hunk 2 above)
Comment (dimitar-dinev) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/AbstractLedgerWorldUpdater.java:148]: That logic has been moved in the [overridden method `HederaStackedWorldStateUpdater`](https://github.com/hashgraph/hedera-services/pull/3152/files#diff-61c1dade85468a15389978645ebb3c22afd195d0ecab11af92df6e0f86e39e42R141-R143)

File: hedera-node/src/main/java/com/hedera/services/store/contracts/AbstractLedgerWorldUpdater.java (hunk 3 above)
Comment (dimitar-dinev) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/AbstractLedgerWorldUpdater.java:160]: That logic has been moved in the [overridden method `HederaStackedWorldStateUpdater`](https://github.com/hashgraph/hedera-services/pull/3152/files#diff-61c1dade85468a15389978645ebb3c22afd195d0ecab11af92df6e0f86e39e42R153-R155)

File: hedera-node/src/main/java/com/hedera/services/store/contracts/HederaStackedWorldStateUpdater.java (hunk 1 above)
Comment (stoqnkpL) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/HederaStackedWorldStateUpdater.java:236]: I think it's better to extract `WorldStateTokenAccount` in its own separate class that has **no** access to state. `WorldStateAccount` should probably remain in `HederaWorldState` since it tightly-coupled to the state held there. My reasoning: 
- we only need to know that the address we're loading is a token address and we should have that info in the current or parent updater, no need to go to the top-level one
- we use the `WorldStateTokenAccount` type to hardcode the redirect bytecode returned by the `getCode()` method, this is a constant one that doesn't need to load from state
- we never want to give access to the world state of a specific updater (especially the top-level one) in a child updater, and creating `WorldStateTokenAccount` object from the context of `HederaWorldState` may leave a backdoor open 
- it would be simpler and easier to read/understand in my opinion

Is there a specific reason not to extract `WorldStateTokenAccount` in it's own class?

Comment (dimitar-dinev) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/HederaStackedWorldStateUpdater.java:236]: Yup, extracting `WorldStateTokenAccount` into a separate class was the initial plan. However, since the [HederaWorldState get() method](https://github.com/hashgraph/hedera-services/blob/f0c1a822b4ced40155074bfd5dd2933ca7dd65ec/hedera-node/src/main/java/com/hedera/services/store/contracts/HederaWorldState.java#L178-L180) (which up until this point has been used for redirecting) returns a `WorldStateAccount`, the newly extracted `WorldStateTokenAccount` has to extend `WorldStateAccount`, but since the latter is an inner class that is not possible 😞

Comment (stoqnkpL) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/HederaStackedWorldStateUpdater.java:236]: Oh, I see... then what about creating new special subtypes of `Account` and `EvmAccount` which encapsulate the hardcoded bytecode + the token address and maybe throw exceptions on any state change attempt?

Comment (dimitar-dinev) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/HederaStackedWorldStateUpdater.java:236]: Hmm, just to make sure we are on the same page --- the idea is to create a new class/classes that will be used **solely** in the stacked updaters and also leave `WorldStateTokenAccount`, right?

Comment (IvanKavaldzhiev) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/HederaStackedWorldStateUpdater.java:236]: What about making WorldStateTokenAccount implements Account, instead of extending MerkleStateAccount. In this case it would be possible to extract WorldStateTokenAccount in a separate class.

Comment (IvanKavaldzhiev) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/HederaStackedWorldStateUpdater.java:236]: In this way we would also be able to extract TOKEN_BYTECODE_PATTERN and TOKEN_CALL_REDIRECT_CONTRACT_BINARY in this new class.

Comment (dimitar-dinev) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/HederaStackedWorldStateUpdater.java:236]: Wouldn't that be impossible again, though, since in this case `WorldStateTokenAccount` wouldn't be extending `WorldStateAccount` and we wouldn't be able to return it from the `get()` in `HederaWorldState` (which returns a `WorldStateAccount`) or am I missing something?

Comment (IvanKavaldzhiev) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/HederaStackedWorldStateUpdater.java:236]: We can explore the option `HederaWorldState.get()` to return `Account` interface. If we get one step back, we can see that `AbstractLedgerWorldUpdate.get()` already returns `Account` interface.

Comment (stoqnkpL) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/HederaStackedWorldStateUpdater.java:236]: I was thinking:
- not changing the behaviour in `HederaWorldState`
- adding new token "stubs" - probably an implementation of `EvmAccount` with implementation for `Bytes getCode()` returning the redirect bytecode and all other methods throwing an exception (we don't expect to call them on a token address AFAICT)
- returning the stub from `AbstractLedgerWorldUpdater` `get(address)` and `getAccount(address)`

Comment (dimitar-dinev) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/HederaStackedWorldStateUpdater.java:236]: That will work, with the expense of introducing some amount of duplication/redundancy... I agree that changing `HederaWorldState` should be treated very carefully and if we were to do it it would require more examination and research  on breaking existing behavior

Comment (stoqnkpL) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/HederaStackedWorldStateUpdater.java:236]: @tinker-michaelj can you advise what you think the best approach to solve this would be in your opinion?
TLDR: is the solution proposed by @dimitar-dinev acceptable, or should we go for a "stubs" approach?

**Review by stoqnkpL — APPROVED** [review]
LGTM!
